django/forms/fields.py

```diff
@@ -1104,17 +1104,20 @@ def __init__(self, path, *, match=None, recursive=False, allow_files=True,
                             f = os.path.join(root, f)
                             self.choices.append((f, f.replace(path, "", 1)))
         else:
+            choices = []
             try:
-                for f in sorted(os.listdir(self.path)):
-                    if f == '__pycache__':
+                for f in os.scandir(self.path):
+                    if f.name == '__pycache__':
                         continue
-                    full_file = os.path.join(self.path, f)
-                    if (((self.allow_files and os.path.isfile(full_file)) or
-                            (self.allow_folders and os.path.isdir(full_file))) and
-                            (self.match is None or self.match_re.search(f))):
-                        self.choices.append((full_file, f))
+                    if (((self.allow_files and f.is_file()) or
+                            (self.allow_folders and f.is_dir())) and
+                            (self.match is None or self.match_re.search(f.name))):
+                        choices.append((f.path, f.name))
+                    choices.sort(key=lambda c: c[1])
             except OSError:
                 pass
```

Comment:
Now that there isn't a potential race condition between calling `listdir` and `isfile`/`isdir`, perhaps this try/except can be removed too.